segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/extendeds3/ExtendedS3Storage.java

```diff
@@ -0,0 +1,384 @@
+        LoggerHelpers.traceLeave(log, "openRead", traceId, streamSegmentName);
+        return retHandle;
+    }
+
+    private SegmentHandle syncOpenWrite(String streamSegmentName) {
+        long traceId = LoggerHelpers.traceEnter(log, "openWrite", streamSegmentName);
+        StreamSegmentInformation info = syncGetStreamSegmentInfo(streamSegmentName);
+        ExtendedS3SegmentHandle retHandle;
+        if (info.isSealed()) {
+            retHandle = ExtendedS3SegmentHandle.getReadHandle(streamSegmentName);
+        } else {
+            retHandle = ExtendedS3SegmentHandle.getWriteHandle(streamSegmentName);
+        }
+
+        LoggerHelpers.traceLeave(log, "openWrite", traceId);
+        return retHandle;
+    }
+
+    private int syncRead(SegmentHandle handle, long offset, byte[] buffer, int bufferOffset, int length) throws IOException, StreamSegmentNotExistsException {
+        long traceId = LoggerHelpers.traceEnter(log, "read", handle.getSegmentName(), offset, bufferOffset, length);
+
+        if (offset < 0 || bufferOffset < 0 || length < 0) {
+            throw new ArrayIndexOutOfBoundsException();
+        }
+
+        try (InputStream reader = client.readObjectStream(config.getBucket(),
+                config.getRoot() + handle.getSegmentName(), Range.fromOffsetLength(offset, length))) {
+
+            if (reader == null) {
+                throw new StreamSegmentNotExistsException(handle.getSegmentName());
+            }
+
+            int bytesRead = StreamHelpers.readAll(reader, buffer, bufferOffset, length);
+
+            LoggerHelpers.traceLeave(log, "read", traceId, bytesRead);
+            return length;
+        }
+    }
+
+    private StreamSegmentInformation syncGetStreamSegmentInfo(String streamSegmentName) {
+        long traceId = LoggerHelpers.traceEnter(log, "getStreamSegmentInfo", streamSegmentName);
+        S3ObjectMetadata result = client.getObjectMetadata(config.getBucket(),
+                config.getRoot() + streamSegmentName);
+
+        AccessControlList acls = client.getObjectAcl(config.getBucket(), config.getRoot() + streamSegmentName);
+
+        boolean canWrite = acls.getGrants().stream().anyMatch((grant) -> grant.getPermission().compareTo(Permission.WRITE) >= 0);
+
+        StreamSegmentInformation information = new StreamSegmentInformation(streamSegmentName,
+                result.getContentLength(), !canWrite, false,
+                new ImmutableDate(result.getLastModified().toInstant().toEpochMilli()));
+        LoggerHelpers.traceLeave(log, "getStreamSegmentInfo", traceId, streamSegmentName);
+        return information;
+    }
+
+    private boolean syncExists(String streamSegmentName) {
+        ListObjectsResult result = null;
+        try {
+            result = client.listObjects(config.getBucket(), config.getRoot() + streamSegmentName);
+            return !result.getObjects().isEmpty();
+        } catch (S3Exception e) {
+            if ( e.getErrorCode().equals("NoSuchKey")) {
```

Comment:
I'm confused about this. Does it throw an exception or return empty in the absence of the requested key?